Mathlib/Computability/TuringDegrees.lean

```diff
@@ -0,0 +1,355 @@
+  case left =>
+    apply Nat.Partrec.left
+  case right =>
+    apply Nat.Partrec.right
+  case oracle g =>
+    rw [← h]
+    apply Nat.Partrec.zero
+  case pair _ _ _ _ ih1 ih2 =>
+    apply Nat.Partrec.pair
+    · apply ih1
+      rw [← h]
+    · apply ih2
+      rw [← h]
+  case comp _ _ _ _ ih1 ih2 =>
+    apply Nat.Partrec.comp
+    · apply ih1
+      rw [← h]
+    · apply ih2
+      rw [← h]
+  case prec _ _ _ _ ih1 ih2 =>
+    apply Nat.Partrec.prec
+    · apply ih1
+      rw [← h]
+    · apply ih2
+      rw [← h]
+  case rfind _ _ ih =>
+    apply Nat.Partrec.rfind
+    apply ih
+    rw [← h]
+
+/--
+A partial function `f` is partial recursive if and only if it is recursive in
+every partial function `g`.
+-/
+theorem partrec_iff_partrec_in_everything
+  (f : ℕ →. ℕ) : Nat.Partrec f ↔ (∀ g, RecursiveIn f g) := by
+  constructor
+  · exact partrec_implies_recursive_in_everything f
+  · intro H
+    have lem : RecursiveIn f (fun _ => Part.some 0) := H (fun _ => Part.some 0)
+    have lem : Nat.Partrec f := partrec_in_zero_implies_partrec f lem
+    exact lem
+
+/--
+Proof that turing reducibility is reflexive.
+-/
+theorem RecursiveIn.refl (f : ℕ →. ℕ) : RecursiveIn f f :=
+  RecursiveIn.oracle
+
+/--
+Proof that `turing_equivalent` is reflexive.
+-/
+@[refl]
+theorem TuringEquivalent.refl (f : ℕ →. ℕ) : f ≡ᵀ f :=
+  ⟨RecursiveIn.refl f, RecursiveIn.refl f⟩
+
+/--
+Proof that `turing_equivalent` is symmetric.
+-/
+@[symm]
+theorem TuringEquivalent.symm {f g : ℕ →. ℕ} (h : f ≡ᵀ g) : g ≡ᵀ f :=
+  ⟨h.2, h.1⟩
+
+/--
+Proof that turing reducibility is transitive.
+-/
+@[trans]
+theorem RecursiveIn.trans {f g h : ℕ →. ℕ} :
+  RecursiveIn f g → RecursiveIn g h → RecursiveIn f h := by
+    intro hg hh
+    induction hg
+    case zero =>
+      apply RecursiveIn.zero
+    case succ =>
+      apply RecursiveIn.succ
+    case left =>
+      apply RecursiveIn.left
+    case right =>
+      apply RecursiveIn.right
+    case oracle =>
+      exact hh
+    case pair f' h' _ _ hf_ih hh_ih =>
+      apply RecursiveIn.pair
+      · apply hf_ih
+        apply hh
+      · apply hh_ih
+        apply hh
+    case comp f' h' _ _ hf_ih hh_ih =>
+      apply RecursiveIn.comp
+      · apply hf_ih
+        apply hh
+      · apply hh_ih
+        apply hh
+    case prec f' h' _ _ hf_ih hh_ih =>
+      apply RecursiveIn.prec
+      · apply hf_ih
+        apply hh
+      · apply hh_ih
+        apply hh
+    case rfind f' _ hf_ih =>
+      apply RecursiveIn.rfind
+      · apply hf_ih
+        apply hh
+
+/--
+Proof that `turing_equivalent` is transitive.
+-/
+theorem TuringEquivalent.trans :
+  Transitive TuringEquivalent :=
+  fun _ _ _ ⟨fg₁, fg₂⟩ ⟨gh₁, gh₂⟩ =>
+    ⟨RecursiveIn.trans fg₁ gh₁, RecursiveIn.trans gh₂ fg₂⟩
+
+/--
+Instance declaring that `turing_equivalent` is an equivalence relation.
+-/
+instance : Equivalence TuringEquivalent :=
+  {
+    refl := TuringEquivalent.refl,
+    symm := TuringEquivalent.symm,
+    trans := @TuringEquivalent.trans
+  }
```

Comment:
If you provide an `IsPreorder` instance for `RecursiveIn` this can be `AntisymmRel.setoid.iseqv`